File: JANL.VB/SQL/ScalarProcedure.vb
Imports System.Data.SqlClient
Imports System.Runtime.CompilerServices
Imports System.Threading.Tasks

Namespace SQL
	''' <summary>
	''' Процедура возвращающая <typeparamref name="T"/>
	''' </summary>
	''' <typeparam name="T">Тип значения возвращаемого процедурой</typeparam>
	Public Class ScalarProcedure(Of T)
		Inherits SQLProcedure(Of T)

		''' <summary>
		''' Создаёт новую процедуру с именем вызывающего метода
		''' </summary>
		Public Sub New(<CallerMemberName> Optional Name As String = Nothing)
			MyBase.New(Name)
		End Sub

		''' <summary>
		''' Выполнить с указанным соединением
		''' </summary>
		Public Overrides Function Execute(Connection As SqlConnection) As T
			SQLCommand.Connection = Connection
			Dim Result = DirectCast(SQLCommand.ExecuteScalar(), T)
			Return Result
		End Function

		''' <summary>
		''' Асинхронно выполнить с указанным соединением
		''' </summary>
		Public Overrides Async Function ExecuteAsync(Connection As SqlConnection) As Task(Of T)
			SQLCommand.Connection = Connection
			Dim Result = DirectCast(Await SQLCommand.ExecuteScalarAsync(), T)
			Return Result
		End Function

	End Class
End Namespace

File: JANL.CS/SQL/ScalarProcedure.cs
using System.Data.SqlClient;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;

namespace JANL.SQL
{
	/// <summary>
	/// Процедура возвращающая <typeparamref name="T"/>
	/// </summary>
	/// <typeparam name="T">Тип значения возвращаемого процедурой</typeparam>
	public class ScalarProcedure<T> : SQLProcedure<T>
	{
		/// <summary>
		/// Создаёт новую процедуру именем вызывающего метода
		/// </summary>
		public ScalarProcedure([CallerMemberName] string Name = null) : base(Name) { }

		/// <summary>
		/// Выполнить с указанным соединением
		/// </summary>
		public override T Execute(SqlConnection Connection)
		{
			SQLCommand.Connection = Connection;
			var Result = (T)SQLCommand.ExecuteScalar();
			return Result;
		}

		/// <summary>
		/// Асинхронно выполнить с указанным соединением
		/// </summary>
		public override async Task<T> ExecuteAsync(SqlConnection Connection)
		{
			SQLCommand.Connection = Connection;
			var Result = (T)await SQLCommand.ExecuteScalarAsync();
			return Result;
		}
	}
}